[GStreamer] Support for bitrate-mode and latency-mode configuration in the video encoder

philn opened the pull request:
#### 634131b13519fc84dee0e194bf1824d7d6ef6034
<pre>
[GStreamer] Support for bitrate-mode and latency-mode configuration in the video encoder
<a href="https://bugs.webkit.org/show_bug.cgi?id=250385">https://bugs.webkit.org/show_bug.cgi?id=250385</a>

Reviewed by Xabier Rodriguez-Calvar.

The video encoder can now be configured to encode at constant (default) or variable bitrate. The
default latency mode is set to realtime, for WebRTC use-cases, but can be set to &quot;quality&quot;, if
mandated by WebCodecs video encoder configuration.

* Source/WebCore/platform/mediastream/gstreamer/GStreamerVideoEncoder.cpp:
(webrtcVideoEncoderBitrateModeGetType):
(webrtcVideoEncoderLatencyModeGetType):
(Encoders::registerEncoder):
(webrtcVideoEncoderGetProperty):
(webrtcVideoEncoderSetProperty):
(webkit_webrtc_video_encoder_class_init):

Canonical link: <a href="https://commits.webkit.org/258825@main">https://commits.webkit.org/258825@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/56a6e225748ea6767762f643744f9e169d7e17c1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103065 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12190 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36084 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112313 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/172524 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/107022 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13209 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3092 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95287 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/110567 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108839 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/10143 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/93336 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/37768 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91981 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/24879 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/79501 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5606 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/26286 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5770 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/2738 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11769 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45788 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6069 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7521 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/platform/mediastream/gstreamer/GStreamerVideoEncoder.cpp:
```diff
@@ -462,21 +546,63 @@ static void webkit_webrtc_video_encoder_class_init(WebKitWebrtcVideoEncoderClass
+                gst_preset_load_preset(GST_PRESET(encoder), "Profile Realtime");
+                break;
+            case QUALITY_LATENCY_MODE:
+                g_object_set(encoder, "threads", 4, "cpu-used", 4, "deadline", 0, "lag-in-frames", 25, nullptr);
```
Where do these 4 threads and 4 cpus come from? Why is it cabled here? Shouldn't this be a parameter somehow? (I mean here and everywhere this is used) Maybe even just a `#define`?

philn replied:
I think that dates back from original code Thibault wrote? idk.

philn replied:
I can add a define for now and let us revisit this later on...

philn committed:
[GStreamer] Support for bitrate-mode and latency-mode configuration in the video encoder
https://bugs.webkit.org/show_bug.cgi?id=250385

Reviewed by Xabier Rodriguez-Calvar.

The video encoder can now be configured to encode at constant (default) or variable bitrate. The
default latency mode is set to realtime, for WebRTC use-cases, but can be set to "quality", if
mandated by WebCodecs video encoder configuration.

* Source/WebCore/platform/mediastream/gstreamer/GStreamerVideoEncoder.cpp:
(webrtcVideoEncoderBitrateModeGetType):
(webrtcVideoEncoderLatencyModeGetType):
(Encoders::registerEncoder):
(webrtcVideoEncoderGetProperty):
(webrtcVideoEncoderSetProperty):
(webkit_webrtc_video_encoder_class_init):

Canonical link: https://commits.webkit.org/258825@main
